### PR TITLE
Feature kitosudv 684 add ef interceptor

### PR DESCRIPTION
**Pull request by mrjsawdk**

### CODE REVIEW

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+        }
+
+        private DbCommandTree HandleUpdateCommand(DbUpdateCommandTree updateCommand)
+        {
+            var userId = GetActiveUserId();
+            var now = _operationClock().Now;
+
+            var updates = new List<KeyValuePair<Predicate<DbSetClause>, DbExpression>>
+            {
+                new KeyValuePair<Predicate<DbSetClause>, DbExpression>(clause=>MatchPropertyName(clause,ObjectOwnerIdColumnName) && MatchNull(clause), userId), //Some EF updates end up in this e.g. changing an owned child on a parent
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:87]: @JacobMalling OBS på denne. Den er lidt speciel. Det sker f.eks. hvor man skifter exhibit ud med en ny c# instans. Her laver den først en update hvor den afkobler den gamle (og hvor man så rammer denne case) og en anden INSERT med den nye instans der patches af HandleCreated... Det er altså lidt en edge case, men vores tests fangede den.

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+                .Select(clause => ApplyUpdates(clause, updates))
+                .ToList();
+
+            return new DbUpdateCommandTree(
+                updateCommand.MetadataWorkspace,
+                updateCommand.DataSpace,
+                updateCommand.Target,
+                updateCommand.Predicate,
+                setClauses.AsReadOnly(),
+                updateCommand.Returning);
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:102]: @JacobMalling  jeg smed Returning fra input på. Den var hard coded til null før og jeg kan godt se at den oftest er null i input, men ingen grund til at miste data hvis det ændrer sig

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+using Infrastructure.Services.Delegates;
+
+namespace Infrastructure.DataAccess.Interceptors
+{
+    public class EFEntityInterceptor : IDbCommandTreeInterceptor
+    {
+        private readonly Factory<IOperationClock> _operationClock;
+        private readonly Factory<Maybe<ActiveUserIdContext>> _userContext;
+        private readonly Factory<IFallbackUserResolver> _fallbackUserResolver;
+        private const string ObjectOwnerIdColumnName = nameof(IEntity.ObjectOwnerId);
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:21]: @JacobMalling OBS på hvordan de her navne konstrueres kontra en konstant string (det her er robust og følger de faktiske navne i modellen)

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+        private const string LastChangedColumnName = nameof(IEntity.LastChanged);
+
+        public EFEntityInterceptor(
+            Factory<IOperationClock> operationClock,
+            Factory<Maybe<ActiveUserIdContext>> userContext,
+            Factory<IFallbackUserResolver> fallbackUserResolver)
+        {
+            _operationClock = operationClock;
+            _userContext = userContext;
+            _fallbackUserResolver = fallbackUserResolver;
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:32]: @JacobMalling OBS: Jeg har tilføjet en FallbackUserResolver. Den er f.eks. interessant ved operationer der ændrer i data når vi ikke har en aktiv bruger. Det er f.eks. ifm. baggrundsjobs. Det er dog sjældent andre steder end lige her vi har dét behov.

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+                    case DbUpdateCommandTree updateCommand:
+                        interceptionContext.Result = HandleUpdateCommand(updateCommand);
+                        return;
+                }
+            }
+        }
+
+        private static bool ShouldHandle(DbCommandTreeInterceptionContext interceptionContext)
+        {
+            return interceptionContext.OriginalResult.DataSpace == DataSpace.SSpace;
```
Comment (mrjsawdk) [inline Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:53]: @JacobMalling Hvorfor SSpace? Kan du uddybe?

Comment (JacobMalling) [reply Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:53]: Jeg er usikker på hvad det gør og har ikke kunne finde information omkring det på nettet. Men alle steder jeg så hvor en interceptor ikke kun blev brugt til logging der havde de den med. Men hvis det ikke er nødvendigt så skal den bare væk

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs (hunk 1 above)
Comment (JacobMalling) [reply Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:87]: Okay så her forsøgte den at ændre på child når den opdaterede parent?

File: Presentation.Web/App_Start/NinjectWebCommon.cs [hunk 6]
```diff
@@ -23,9 +30,17 @@ public static void Start()
...
 
             //Setup for web application
-            bootstrapper.Initialize(()=> new KernelBuilder().ForWebApplication().Build());
+            bootstrapper.Initialize(() =>
+            {
+                var kernel = new KernelBuilder().ForWebApplication().Build();
 
-            //Setup for hangfire with a different setup
+                //Only register the interceptor once per application (Hangfire job is the same application scope)
+                DbInterception.Add(new EFEntityInterceptor(() => kernel.Get<IOperationClock>(), () => kernel.Get<Maybe<ActiveUserIdContext>>(), () => kernel.Get<IFallbackUserResolver>()));
```
Comment (JacobMalling) [inline Presentation.Web/App_Start/NinjectWebCommon.cs:38]: Ahh så på denne måde får du aktiveret ninject kernel og resolvet de ting der måtte være der og så derefter tilføjet interceptoren. Så er jeg glad for jeg smed denne opgave tilbage da de løsninger jeg var ved at overveje var noget forsøg med lazy loading

File: Core.DomainServices/Context/ActiveUserIdContext.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Core.DomainServices.Context
+{
+    public class ActiveUserIdContext
+    {
+        public int ActiveUserId { get; }
+
+        public ActiveUserIdContext(int activeUserId)
+        {
+            ActiveUserId = activeUserId;
```
[the remark this answers is not in the document]
Comment (JacobMalling) [reply Core.DomainServices/Context/ActiveUserIdContext.cs:9]: Kan godt se vi ikke behøver hele usercontext, men er det den eneste grund? eller er der også noget i forhold til dependency resolve?

[the remark this answers is not in the document]
Comment (mrjsawdk) [reply Core.DomainServices/Context/ActiveUserIdContext.cs:9]: Det er for at man kan få user id uden at skulle vide mere :-)

[the remark this answers is not in the document]
Comment (mrjsawdk) [reply Core.DomainServices/Context/ActiveUserIdContext.cs:9]: Og den er registreret som Maybe

File: Presentation.Web/App_Start/NinjectWebCommon.cs (hunk 6 above)
Comment (mrjsawdk) [reply Presentation.Web/App_Start/NinjectWebCommon.cs:38]: Ja det der interception er ikke integreret i almindelig pipeline, så vi er bare heldige at Ninject kan finde den rette context når den resolver i kontekst af hhv. request og/eller hangfire tråd

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs (hunk 1 above)
Comment (mrjsawdk) [reply Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:87]: Nej problemet er at selvom "child" er et "child" i C# verdenen så er den jo sin egen tabel i SQL og når man laver de der udskiftningstricks, så kommer der nogle gange flere "validerings-loops" indtil den beslutter sig.. det er i hvert fald min teori for i databasen ser det korrekt ud bagefter :-)

File: Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs (hunk 5 above)
Comment (mrjsawdk) [reply Infrastructure.DataAccess/Interceptors/EFEntityInterceptor.cs:53]: tja den kalder jo TreeCreated flere gange og SSpace er det sidste. Det lader til at virke lige nu, så jeg piller ikke i det
